src/huggingface_hub/hf_file_system.py

```diff
@@ -249,54 +249,124 @@ def ls(
+            path = f"{self._api.endpoint}/api/{resolved_path.repo_type}s/{resolved_path.repo_id}/tree/{safe_quote(resolved_path.revision)}/{resolved_path.path_in_repo}".rstrip(
+                "/"
+            )
+            headers = self._api._build_hf_headers()
+            params = {"recursive": recursive} if recursive else {}
+            yield from paginate(path, params=params, headers=headers)
+
+        def _format_tree_item_as_path_info(tree_item: Dict) -> Dict:
+            path_info = {
+                "name": path_prefix + tree_item["path"],
+                "size": tree_item["size"],
+                "type": tree_item["type"],
+            }
+            if tree_item["type"] == "file":
+                path_info.update(
+                    {
+                        "blob_id": tree_item["oid"],
+                        "lfs": tree_item.get("lfs"),
+                        "last_modified": parse_datetime(tree_item["lastCommit"]["date"]),
+                    },
+                )
+            return path_info
+
+        out = []
+        if path not in self.dircache or refresh:
+            for tree_item in _iter_repo_tree(resolved_path, recursive=recursive):
+                path_info = _format_tree_item_as_path_info(tree_item)
+                parent_path = (path_prefix + self._parent(tree_item["path"])).rstrip("/")
+                self.dircache.setdefault(parent_path, []).append(path_info)
+                out.append(path_info)
+        else:
+            cached_infos = self.dircache[path]
+            out.extend(cached_infos)
+            # Use BFS to traverse the cache to build the "recursive "output
+            # - if the cache is incomplete, find the common prefix of the missing entries and extend the output with the result of `_ls_tree(common_prefix, recursive=True)`
+            if recursive:
+                dirs_to_visit = deque([path_info for path_info in cached_infos if path_info["type"] == "directory"])
+                dirs_not_in_cache = []
+                while dirs_to_visit:
+                    dir_info = dirs_to_visit.popleft()
+                    if dir_info["name"] not in self.dircache:
+                        dirs_not_in_cache.append(dir_info["name"])
+                    else:
+                        cached_infos = self.dircache[dir_info["name"]]
+                        out.extend(cached_infos)
+                        dirs_to_visit.extend(
+                            [path_info for path_info in cached_infos if path_info["type"] == "directory"]
+                        )
+                if dirs_not_in_cache:
+                    dirs_not_in_cache = [dir_path[len(path_prefix) :] for dir_path in dirs_not_in_cache]
+                    common_path = (path_prefix + os.path.commonpath(dirs_not_in_cache)).rstrip("/")
+                    out = [o for o in out if not o["name"].startswith(common_path)]
+                    self.dircache.pop(common_path, None)
```

Comment:
Can you add a comment to explain why you need to remove the common path cache here ?

Also what if the common path was already visited, would we end up with duplicates ?

Reply:
Sure, I can add a comment.

Yes, it's precisely why we need this part - to avoid duplicates if the common path was already visited.